### PR TITLE
more util functions to the arrays module

### PR DESCRIPTION
**Pull request by prashanth-hegde**
Adds more utility functions to the arrays module, in the spirit of making working with arrays more pleasant

`find_first` - finds first element matching a predicate 
`find_last` - finds last element matching a predicate 
`all_of` - true if all elements match a predicate
`any_of` - true if any element matches a predicate
`join_to_string` - uses a transform function to stringify and joins them all with a separator

I used `any_of()` instead of `any()` because `any` seemed to conflict with a keyword. 
And likewise `all_of()` to be consistent

### <samp>🤖 Generated by Copilot at b614038</samp>

This pull request enhances the `arrays` module with new functions for common array operations and adds unit tests for them. The new functions are `find_first`, `find_last`, `join_to_string`, `any_of`, and `all_of`, and they work with arrays of any type. The unit tests use the `arrays_test.v` file and cover different scenarios and edge cases.

### <samp>🤖 Generated by Copilot at b614038</samp>

*  Add five new functions to the arrays module for searching, filtering, and transforming arrays of any type ([link](https://github.com/vlang/v/pull/18784/files?diff=unified&w=0#diff-b7b2841a2601a108289f795e922710054cff56ff06212c11d40dd2a0346191dcR667-R727))
* Add unit tests for the new functions in the arrays module using `assert` and `or` blocks ([link](https://github.com/vlang/v/pull/18784/files?diff=unified&w=0#diff-41085077bf9d89c78db1526d74fcf1fce586b5c1e4116f467f08330d6b471da5R407-R471))

### CODE REVIEW

File: vlib/arrays/arrays.v [hunk 1]
```diff
@@ -664,3 +664,64 @@ pub fn carray_to_varray[T](c_array_data voidptr, items int) []T {
...
+  }
+  for item in array {
+    if predicate(item) {
+      return item
+    }
+  }
+  return none
+}
+
+// find_last returns the first element that matches the given predicate
```
Comment (Larpon) [inline vlib/arrays/arrays.v:683]:
```suggestion
// find_last returns the last element that matches the given predicate
```

**Review by Larpon — CHANGES_REQUESTED** [review]

File: vlib/arrays/arrays.v (hunk 1 above)
Comment (prashanth-hegde) [reply vlib/arrays/arrays.v]: good catch, Thank you

File: vlib/arrays/arrays.v [hunk 2]
```diff
@@ -664,3 +664,64 @@ pub fn carray_to_varray[T](c_array_data voidptr, items int) []T {
...
+pub fn join_to_string[T](array []T, separator string, transform fn (elem T) string) string {
+  mut builder := []string{len: array.len, cap: array.len}
+  for i, item in array {
+    builder[i] = transform(item)
+  }
+  return builder.join(separator)
+}
+
+// any_of returns true if any of the elements in the array matches the predicate
+pub fn any_of[T](array []T, predicate fn (elem T) bool) bool {
```
Comment (spytheman) [inline vlib/arrays/arrays.v:710]: What is the difference with just using the builtin `a.any(predicate)` ?
Try for example: `println( [1,2,3].any(it==5) )` or `println( [9,2,3].any(it==9) )` .

File: vlib/arrays/arrays.v [hunk 3]
```diff
@@ -664,3 +664,64 @@ pub fn carray_to_varray[T](c_array_data voidptr, items int) []T {
...
+  for item in array {
+    if predicate(item) {
+      return true
+    }
+  }
+  return false
+}
+
+// all_of returns true if all of the elements in the array matches the predicate
+pub fn all_of[T](array []T, predicate fn (elem T) bool) bool {
```
Comment (spytheman) [inline vlib/arrays/arrays.v:720]: What is the difference with just using the builtin method `a.all(predicate)` ?
Try for example `println( [1,2,3].all(it<9) )`

File: vlib/arrays/arrays.v (hunk 2 above)
Comment (prashanth-hegde) [reply vlib/arrays/arrays.v:710]: good call. I somehow missed it in the docs. Thanks for pointing out. I will remove this function. Same with the `all()` function also

File: vlib/arrays/arrays.v [hunk 4]
```diff
@@ -664,3 +664,45 @@ pub fn carray_to_varray[T](c_array_data voidptr, items int) []T {
 	unsafe { vmemcpy(v_array.data, c_array_data, total_size) }
 	return v_array
 }
+
+// find_first returns the first element that matches the given predicate
+// returns an optional in case none of the elements match the predicate
```
Comment (spytheman) [inline vlib/arrays/arrays.v:669]:
```suggestion
// returns `none`, if there is no match found
```

File: vlib/arrays/arrays.v [hunk 5]
```diff
@@ -664,3 +664,45 @@ pub fn carray_to_varray[T](c_array_data voidptr, items int) []T {
...
+      return item
+    }
+  }
+  return none
+}
+
+// join_to_string takes in a custom transform function and joins all elements into a string with
+// the specified separator
+pub fn join_to_string[T](array []T, separator string, transform fn (elem T) string) string {
+  mut builder := []string{len: array.len, cap: array.len}
```
Comment (spytheman) [inline vlib/arrays/arrays.v:702]:
```suggestion
  mut builder := []string{len: array.len}
```

Comment (spytheman) [reply vlib/arrays/arrays.v:702]: setting cap is done internally too, by setting len

File: vlib/arrays/arrays_test.v [hunk 6]
```diff
@@ -404,3 +404,59 @@ fn test_map_of_counts() {
...
+  })?
+  assert find_by_name == FindTest{'one', 1}
+
+  // not found
+  find_none := find_first(test_structs, fn (arr FindTest) bool {
+    return arr.name == 'nothing'
+  }) or {
+    assert true
+    return
+  }
```
Comment (spytheman) [inline vlib/arrays/arrays_test.v:432]:
```suggestion
  }
  assert false
```

Comment (spytheman) [reply vlib/arrays/arrays_test.v:432]: alternatively, use `if _ := f() { assert false } else { assert true }`

File: vlib/arrays/arrays_test.v [hunk 7]
```diff
@@ -404,3 +404,59 @@ fn test_map_of_counts() {
...
+  })?
+  assert find_by_name == FindTest{'one', 4}
+
+  // not found
+  find_none := find_last(test_structs, fn (arr FindTest) bool {
+    return arr.name == 'nothing'
+  }) or {
+    assert true
+    return
+  }
```
Comment (spytheman) [inline vlib/arrays/arrays_test.v:454]:
```suggestion
  }
  assert false
```

File: vlib/arrays/arrays_test.v (hunk 6 above)
Comment (prashanth-hegde) [reply vlib/arrays/arrays_test.v:432]: when the item is not found, it's an expected result. Because `arr.name == 'nothing'` Hence the `or {}` clause has assert true

Comment (spytheman) [reply vlib/arrays/arrays_test.v]: That is what the `if _ := f() { assert false } else { assert true }` will do as well - the succeeding assertion is in the else {} branch.

It also assures, that if for some reason, there is a regression in the implementation, the assertion will fail (in the then {} branch).

File: vlib/arrays/arrays.v (hunk 5 above)
Comment (spytheman) [reply vlib/arrays/arrays.v]: In fact the function can be done like this, to be more efficient (it avoids allocating the intermediate array):
```v
import strings
...
pub fn join_to_string[T](array []T, separator string, transform fn (elem T) string) string {
    mut sb := strings.new_builder(array.len*2)
    defer {
        unsafe { sb.free() }
    }
    for i, item in array {
        sb.write_string(transform(item))
        if i < array.len - 1 {
            sb.write_string(separator)
        }
    }
    return sb.str()
}
```

File: vlib/arrays/arrays_test.v (hunk 6 above)
Comment (prashanth-hegde) [reply vlib/arrays/arrays_test.v]: fair point. updated the tests

File: vlib/arrays/arrays.v (hunk 5 above)
Comment (spytheman) [reply vlib/arrays/arrays.v]: The previous version is more or less equivalent to `a.map(transform).join(separator)` .

Comment (prashanth-hegde) [reply vlib/arrays/arrays.v]: so should I just use `a.map(transform).join(separator)` instead?

Comment (spytheman) [reply vlib/arrays/arrays.v]: No. With the new version, it is more efficient and less memory intensive, compared to `a.map(transform).join(separator)`, since it skips one intermediate step (creating a temporary []string, just so right after that, it can be used with `.join()`.
